bootstrapping fails on Fedora 43

### Steps to reproduce

```console
docker run --rm -it fedora:43
yum install -y gcc g++ python3 cmake git
git clone --depth=1 https://github.com/spack/spack
alias spack='/spack/bin/spack'
git clone --depth=1 https://github.com/spack/spack-packages
spack repo set --destination /spack-packages builtin
spack compiler find
    ==> Added 1 new compiler to /root/.spack/packages.yaml
        gcc@15.2.1
spack bootstrap now
```

I found no difference in using `spack repo set --destination /spack-packages builtin` versus letting spack grab its own version of `spack-packages`.

### Error message

On Fedora 43, the boostrapping fails:

```
spack bootstrap now
==> Fetching https://ghcr.io/v2/spack/bootstrap-buildcache-v1/blobs/sha256:16f27dafb233a9aff14adff75112d0a8b2e03492f796f12101785c02950aa7f6
==> Fetching https://ghcr.io/v2/spack/bootstrap-buildcache-v1/blobs/sha256:b76a4eaef54b24d0ea9b8dfa9392f7ab519f918ae7e1a8bb919539d9adeddbcb
==> Installing "gcc-runtime@=10.2.1 build_system=generic platform=linux os=centos7 target=x86_64" from a buildcache
==> Fetching https://ghcr.io/v2/spack/bootstrap-buildcache-v1/blobs/sha256:5367a0d038a87532fbbe373da31502bd32e399cc113343f403ebbc9d8ca7d552
==> Fetching https://ghcr.io/v2/spack/bootstrap-buildcache-v1/blobs/sha256:79dfb7064e7993a97474c5f6b7560254fe19465a6c4cfc44569852e5a6ab542b
==> Installing "patchelf@=0.17.2 build_system=autotools platform=linux os=centos7 target=x86_64" from a buildcache
==> Warning: Fetching an index from a v2 binary mirror layout, located at
  https://mirror.spack.io/bootstrap/github-actions/v0.5 is deprecated.
  Support for this will be removed in a future version of spack. If you
  manage the buildcache please consider running:
    'spack buildcache migrate'
  or rebuilding the specs in this mirror. Otherwise, consider running:
    'spack mirror list'
    'spack mirror remove <name>'
  with the <name> for the mirror url shown in the list.
==> Installing compiler-wrapper-1.0-jpnthhggz2ur6miccsreuf6p6lnlsvdp [1/15]
==> Fetching https://mirror.spack.io/_source-cache/archive/a5/a5ff4fcdbeda284a7993b87f294b6338434cffc84ced31e4d04008ed5ea389bf
    [100%]   30.08 KB @  125.7 MB/s
==> No patches needed for compiler-wrapper
==> compiler-wrapper: Executing phase: 'install'
==> compiler-wrapper: Successfully installed compiler-wrapper-1.0-jpnthhggz2ur6miccsreuf6p6lnlsvdp
  Stage: 0.04s.  Install: 0.00s.  Post-install: 0.02s.  Total: 0.25s
[+] /root/.spack/bootstrap/store/linux-x86_64/compiler-wrapper-1.0-jpnthhggz2ur6miccsreuf6p6lnlsvdp
[+] /usr (external gcc-15.2.1-6fo7fdrew5dxqj7j5r7rkc44l4hv7rgn)
[+] /usr (external glibc-2.42-hzlxo5ujmypfjfn7ungkrylwbkrblqf6)
[+] /usr (external cmake-3.31.6-fj5r5iy5vber3o2iymigf3r4td674t7e)
[+] /usr (external python-3.14-kmffuc7axp7vka3rbjnmxsch4u5ocr6l)
==> Installing gcc-runtime-15.2.1-qdgouy2d5hfjwt3jap4xs7h4ixefvc3p [6/15]
==> No patches needed for gcc-runtime
==> gcc-runtime: Executing phase: 'install'
==> gcc-runtime: Successfully installed gcc-runtime-15.2.1-qdgouy2d5hfjwt3jap4xs7h4ixefvc3p
  Stage: 0.00s.  Install: 0.03s.  Post-install: 0.01s.  Total: 0.21s
[+] /root/.spack/bootstrap/store/linux-x86_64/gcc-runtime-15.2.1-qdgouy2d5hfjwt3jap4xs7h4ixefvc3p
==> Installing python-venv-1.0-tofejayvwwuojfxpfi6hjmrofsrm73qm [7/15]
==> No patches needed for python-venv
==> python-venv: Executing phase: 'install'
==> python-venv: Successfully installed python-venv-1.0-tofejayvwwuojfxpfi6hjmrofsrm73qm
  Stage: 0.00s.  Install: 0.04s.  Post-install: 0.01s.  Total: 0.20s
[+] /root/.spack/bootstrap/store/linux-x86_64/python-venv-1.0-tofejayvwwuojfxpfi6hjmrofsrm73qm
==> Installing gmake-4.4.1-vfjjpqsvduetcmivo35gsginbdmbug3y [8/15]
==> Fetching https://mirror.spack.io/_source-cache/archive/dd/dd16fb1d67bfab79a72f5e8390735c49e3e8e70b4945a15ab1f81ddb78658fb3.tar.gz
    [100%]    2.35 MB @   49.1 MB/s
==> No patches needed for gmake
==> gmake: Executing phase: 'install'
==> gmake: Successfully installed gmake-4.4.1-vfjjpqsvduetcmivo35gsginbdmbug3y
  Stage: 0.13s.  Install: 10.36s.  Post-install: 0.01s.  Total: 10.55s
[+] /root/.spack/bootstrap/store/linux-x86_64/gmake-4.4.1-vfjjpqsvduetcmivo35gsginbdmbug3y
==> Installing libsigsegv-2.14-er5zdzlqsevy72gapj5tshdk66c2i7qo [9/15]
==> Fetching https://mirror.spack.io/_source-cache/archive/cd/cdac3941803364cf81a908499beb79c200ead60b6b5b40cad124fd1e06caa295.tar.gz
    [100%]  464.66 KB @   21.2 MB/s
==> No patches needed for libsigsegv
==> libsigsegv: Executing phase: 'autoreconf'
==> libsigsegv: Executing phase: 'configure'
==> libsigsegv: Executing phase: 'build'
==> libsigsegv: Executing phase: 'install'
==> libsigsegv: Successfully installed libsigsegv-2.14-er5zdzlqsevy72gapj5tshdk66c2i7qo
  Stage: 0.07s.  Autoreconf: 0.00s.  Configure: 3.48s.  Build: 0.42s.  Install: 0.07s.  Post-install: 0.01s.  Total: 4.32s
[+] /root/.spack/bootstrap/store/linux-x86_64/libsigsegv-2.14-er5zdzlqsevy72gapj5tshdk66c2i7qo
==> Installing re2c-3.1-q4d6it46zbpataqrl3itjbhb5zysh3vm [10/15]
==> Fetching https://mirror.spack.io/_source-cache/archive/0a/0ac299ad359e3f512b06a99397d025cfff81d3be34464ded0656f8a96676c029.tar.xz
    [100%]    1.45 MB @   50.0 MB/s
==> No patches needed for re2c
==> re2c: Executing phase: 'autoreconf'
==> re2c: Executing phase: 'configure'
==> re2c: Executing phase: 'build'
==> re2c: Executing phase: 'install'
==> re2c: Successfully installed re2c-3.1-q4d6it46zbpataqrl3itjbhb5zysh3vm
  Stage: 0.22s.  Autoreconf: 0.00s.  Configure: 3.29s.  Build: 12.92s.  Install: 0.29s.  Post-install: 0.05s.  Total: 16.98s
[+] /root/.spack/bootstrap/store/linux-x86_64/re2c-3.1-q4d6it46zbpataqrl3itjbhb5zysh3vm
==> Installing libiconv-1.17-bw6bykc4xxou4fedsnhnhn36zamkcb2d [11/15]
==> Fetching https://mirror.spack.io/_source-cache/archive/8f/8f74213b56238c85a50a5329f77e06198771e70dd9a739779f4c02f65d971313.tar.gz
    [100%]    5.41 MB @   72.8 MB/s
==> No patches needed for libiconv
==> libiconv: Executing phase: 'autoreconf'
==> libiconv: Executing phase: 'configure'
==> libiconv: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    '/root/.spack/bootstrap/store/linux-x86_64/gmake-4.4.1-vfjjpqsvduetcmivo35gsginbdmbug3y/bin/make' '-j16' 'V=1'

4 errors found in build log:
     453    libtool: compile:  /root/.spack/bootstrap/store/linux-x86_64/compiler-wrapper-1.0-jpnthhggz2ur6miccsreuf6p6lnlsvdp/libexec/spack/gcc/gcc -I. -I. 
            -I../include -I./../include -I.. -I./.. -g -O2 -fvisibility=hidden -DLIBDIR=\"/root/.spack/bootstrap/store/linux-x86_64/libiconv-1.17-bw6bykc4xxo
            u4fedsnhnhn36zamkcb2d/lib\" -DBUILDING_LIBICONV -DBUILDING_LIBCHARSET -DBUILDING_DLL -DENABLE_RELOCATABLE=1 -DIN_LIBRARY -DINSTALLDIR=\"/root/.sp
            ack/bootstrap/store/linux-x86_64/libiconv-1.17-bw6bykc4xxou4fedsnhnhn36zamkcb2d/lib\" -DNO_XMALLOC -Dset_relocation_prefix=libiconv_set_relocatio
            n_prefix -Drelocate=libiconv_relocate -Drelocate2=libiconv_relocate2 -DHAVE_CONFIG_H -c ./../libcharset/lib/localcharset.c  -fPIC -DPIC -o .libs/
            localcharset.o
     454    libtool: compile:  /root/.spack/bootstrap/store/linux-x86_64/compiler-wrapper-1.0-jpnthhggz2ur6miccsreuf6p6lnlsvdp/libexec/spack/gcc/gcc -I. -I. 
            -I../include -I./../include -I.. -I./.. -g -O2 -fvisibility=hidden -DLIBDIR=\"/root/.spack/bootstrap/store/linux-x86_64/libiconv-1.17-bw6bykc4xxo
            u4fedsnhnhn36zamkcb2d/lib\" -DBUILDING_LIBICONV -DBUILDING_LIBCHARSET -DBUILDING_DLL -DENABLE_RELOCATABLE=1 -DIN_LIBRARY -DINSTALLDIR=\"/root/.sp
            ack/bootstrap/store/linux-x86_64/libiconv-1.17-bw6bykc4xxou4fedsnhnhn36zamkcb2d/lib\" -DNO_XMALLOC -Dset_relocation_prefix=libiconv_set_relocatio
            n_prefix -Drelocate=libiconv_relocate -Drelocate2=libiconv_relocate2 -DHAVE_CONFIG_H -c ./relocatable.c  -fPIC -DPIC -o .libs/relocatable.o
     455    libtool: compile:  /root/.spack/bootstrap/store/linux-x86_64/compiler-wrapper-1.0-jpnthhggz2ur6miccsreuf6p6lnlsvdp/libexec/spack/gcc/gcc -I. -I. 
            -I../include -I./../include -I.. -I./.. -g -O2 -fvisibility=hidden -DLIBDIR=\"/root/.spack/bootstrap/store/linux-x86_64/libiconv-1.17-bw6bykc4xxo
            u4fedsnhnhn36zamkcb2d/lib\" -DBUILDING_LIBICONV -DBUILDING_LIBCHARSET -DBUILDING_DLL -DENABLE_RELOCATABLE=1 -DIN_LIBRARY -DINSTALLDIR=\"/root/.sp
            ack/bootstrap/store/linux-x86_64/libiconv-1.17-bw6bykc4xxou4fedsnhnhn36zamkcb2d/lib\" -DNO_XMALLOC -Dset_relocation_prefix=libiconv_set_relocatio
            n_prefix -Drelocate=libiconv_relocate -Drelocate2=libiconv_relocate2 -DHAVE_CONFIG_H -c ./../libcharset/lib/localcharset.c  -fPIC -DPIC -o localc
            harset.o >/dev/null 2>&1
     456    libtool: compile:  /root/.spack/bootstrap/store/linux-x86_64/compiler-wrapper-1.0-jpnthhggz2ur6miccsreuf6p6lnlsvdp/libexec/spack/gcc/gcc -I. -I. 
            -I../include -I./../include -I.. -I./.. -g -O2 -fvisibility=hidden -DLIBDIR=\"/root/.spack/bootstrap/store/linux-x86_64/libiconv-1.17-bw6bykc4xxo
            u4fedsnhnhn36zamkcb2d/lib\" -DBUILDING_LIBICONV -DBUILDING_LIBCHARSET -DBUILDING_DLL -DENABLE_RELOCATABLE=1 -DIN_LIBRARY -DINSTALLDIR=\"/root/.sp
            ack/bootstrap/store/linux-x86_64/libiconv-1.17-bw6bykc4xxou4fedsnhnhn36zamkcb2d/lib\" -DNO_XMALLOC -Dset_relocation_prefix=libiconv_set_relocatio
            n_prefix -Drelocate=libiconv_relocate -Drelocate2=libiconv_relocate2 -DHAVE_CONFIG_H -c ./relocatable.c  -fPIC -DPIC -o relocatable.o >/dev/null 
            2>&1
     457    In file included from ./loops.h:23,
     458                     from ./iconv.c:148:
  >> 459    ./loop_wchar.h:39:17: error: conflicting types for 'mbrtowc'; have 'size_t(void)' {aka 'long unsigned int(void)'}
     460       39 |   extern size_t mbrtowc ();
     461          |                 ^~~~~~~
```
This bug was fixed in https://github.com/spack/spack-packages/pull/1616. It seems I forgot to test it in my Fedora environments... In any case, I cannot figure out why it says there are no patches needed:

```
==> Installing libiconv-1.17-bw6bykc4xxou4fedsnhnhn36zamkcb2d [11/15]
==> Fetching https://mirror.spack.io/_source-cache/archive/8f/8f74213b56238c85a50a5329f77e06198771e70dd9a739779f4c02f65d971313.tar.gz
    [100%]    5.41 MB @   72.8 MB/s
==> No patches needed for libiconv
```

I've checked the files from the downloaded archive that are touched by the `loop_wchar_9eb508.patch` patch, and they do need to be modified. The `patch` configuration also looks correct in the package:

```python
    patch(
        "loop_wchar_9eb508.patch",
        when="@:1.17",
        sha256="e0145a14e9fed1d9f07003a4772db916faa8cbb7c5273880a38d7c2e64d4103c",
    )
```

I am at a loss. Thoughts are very much welcome.

---

This issue doesn't happen on Ubuntu 25.10 that also has gcc-15 because all of the bootstrap binaries are downloaded from cache.

```console
docker run --rm -it ubuntu:25.10
apt update -y
apt install -y build-essential gcc g++ python3 cmake ca-certificates git
git clone --depth=1 https://github.com/spack/spack
alias spack='/spack/bin/spack'
git clone --depth=1 https://github.com/spack/spack-packages
spack repo set --destination /spack-packages builtin
spack compiler find
    ==> Added 1 new compiler to /root/.spack/packages.yaml
        gcc@15.2.0
spack bootstrap now
    ==> Fetching https://ghcr.io/v2/spack/bootstrap-buildcache-v1/blobs/sha256:16f27dafb233a9aff14adff75112d0a8b2e03492f796f12101785c02950aa7f6
    ==> Fetching https://ghcr.io/v2/spack/bootstrap-buildcache-v1/blobs/sha256:b76a4eaef54b24d0ea9b8dfa9392f7ab519f918ae7e1a8bb919539d9adeddbcb
    ==> Installing "gcc-runtime@=10.2.1 build_system=generic platform=linux os=centos7 target=x86_64" from a buildcache
    ==> Fetching https://ghcr.io/v2/spack/bootstrap-buildcache-v1/blobs/sha256:5367a0d038a87532fbbe373da31502bd32e399cc113343f403ebbc9d8ca7d552
    ==> Fetching https://ghcr.io/v2/spack/bootstrap-buildcache-v1/blobs/sha256:79dfb7064e7993a97474c5f6b7560254fe19465a6c4cfc44569852e5a6ab542b
    ==> Installing "patchelf@=0.17.2 build_system=autotools platform=linux os=centos7 target=x86_64" from a buildcache
    ==> Fetching https://ghcr.io/v2/spack/bootstrap-buildcache-v1/blobs/sha256:9ac106d746fc9830acbb2e6b135a5c1748482df8fe115bee2deb1ed02e5189c4
    ==> Fetching https://ghcr.io/v2/spack/bootstrap-buildcache-v1/blobs/sha256:6599ac06ade0cb3e80695f36492ea94a306f8bde0537482521510076c5981aa0
    ==> Installing "clingo-bootstrap@=spack~docs+ipo+optimized+python+static_libstdcpp build_system=cmake build_type=Release generator=make patches:=bebb819,ec99431 platform=linux os=centos7 target=x86_64" from a buildcache
spack find
    ==> 0 installed packages
```

### Information on your system

```
spack debug report
* **Spack:** 1.1.0.dev0 (https://github.com/spack/spack/commit/7554d2a4a3bd3ac67075277cc34c1fd02193e0dd)
* **Builtin repo:** https://github.com/spack/spack-packages/commit/7898974e2be1df0f4497747eac2581b74e32ef2e
* **Python:** 3.14.0rc3
* **Platform:** linux-fedora43-zen3
```

### General information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have searched the issues of this repo and believe this is not a duplicate
- [x] I have run the failing commands in debug mode and reported the output

--- a/lib/spack/spack/bootstrap/clingo.py
+++ b/lib/spack/spack/bootstrap/clingo.py
@@ -11,19 +11,30 @@ JSON file for a similar platform.
 """
 import pathlib
 import sys
-from typing import Dict, Optional, Tuple
+from typing import Dict, Optional, Tuple, Type
 
 import spack.vendor.archspec.cpu
 
 import spack.compilers.config
 import spack.compilers.libraries
 import spack.config
+import spack.package_base
 import spack.platforms
+import spack.repo
 import spack.spec
 import spack.traverse
 import spack.version
 
 from .config import spec_for_current_python
+
+
+def _select_best_version(pkg_cls: Type["spack.package_base.PackageBase"], node: spack.spec.Spec):
+    """Try to attach the best known version to a node"""
+    try:
+        best_version = spack.package_base.preferred_version(pkg_cls)
+    except KeyError:
+        return
+    node.versions.versions = [spack.version.from_string(f"={best_version}")]
 
 
 class ClingoBootstrapConcretizer:
@@ -119,13 +130,32 @@ class ClingoBootstrapConcretizer:
         s = spack.spec.Spec.from_specfile(str(self.prototype_path()))
         s._mark_concrete(False)
 
-        # Tweak it to conform to the host architecture
+        # These are nodes in the cmake stack, whose versions are frequently deprecated for
+        # security reasons. In case there is no external cmake on this machine, we'll update
+        # their versions to the most preferred according to the repository we know.
+        to_be_updated = {
+            pkg_name: spack.repo.PATH.get_pkg_class(pkg_name)
+            for pkg_name in [
+                "ca-certificates-mozilla",
+                "openssl",
+                "curl",
+                "cmake",
+                "libiconv",
+                "ncurses",
+            ]
+        }
+
+        # Tweak it to conform to the host architecture + update the version of a few dependencies
         for node in s.traverse():
             node.architecture.os = str(self.host_os)
             node.architecture = self.host_architecture
 
             if node.name == "gcc-runtime":
                 node.versions = self.host_compiler.versions
+
+            if node.name in to_be_updated:
+                pkg_cls = to_be_updated[node.name]
+                _select_best_version(pkg_cls=pkg_cls, node=node)
 
         # Can't use re2c@3.1 with Python 3.6
         if self.host_python.satisfies("@3.6"):

--- a/lib/spack/spack/bootstrap/clingo.py
+++ b/lib/spack/spack/bootstrap/clingo.py
@@ -21,6 +21,7 @@ import spack.config
 import spack.package_base
 import spack.platforms
 import spack.repo
+import spack.solver.versions
 import spack.spec
 import spack.traverse
 import spack.version
@@ -28,11 +29,19 @@ import spack.version
 from .config import spec_for_current_python
 
 
-def _select_best_version(pkg_cls: Type["spack.package_base.PackageBase"], node: spack.spec.Spec):
+def _select_best_version(
+    pkg_cls: Type["spack.package_base.PackageBase"], node: spack.spec.Spec, valid_versions: str
+) -> None:
     """Try to attach the best known version to a node"""
+    constraint = spack.version.from_string(valid_versions)
+    allowed_versions = [
+        (v, info) for v, info in pkg_cls.versions.items() if v.satisfies(constraint)
+    ]
     try:
-        best_version = spack.package_base.preferred_version(pkg_cls)
-    except KeyError:
+        best_version, _ = max(
+            allowed_versions, key=spack.solver.versions.concretization_version_order
+        )
+    except (KeyError, ValueError):
         return
     node.versions.versions = [spack.version.from_string(f"={best_version}")]
 
@@ -132,17 +141,18 @@ class ClingoBootstrapConcretizer:
 
         # These are nodes in the cmake stack, whose versions are frequently deprecated for
         # security reasons. In case there is no external cmake on this machine, we'll update
-        # their versions to the most preferred according to the repository we know.
+        # their versions to the most preferred, within the valid range, according to the
+        # repository we know.
         to_be_updated = {
-            pkg_name: spack.repo.PATH.get_pkg_class(pkg_name)
-            for pkg_name in [
-                "ca-certificates-mozilla",
-                "openssl",
-                "curl",
-                "cmake",
-                "libiconv",
-                "ncurses",
-            ]
+            pkg_name: (spack.repo.PATH.get_pkg_class(pkg_name), valid_versions)
+            for pkg_name, valid_versions in {
+                "ca-certificates-mozilla": ":",
+                "openssl": "3:3",
+                "curl": "8:8",
+                "cmake": "3.16:3",
+                "libiconv": "1:1",
+                "ncurses": "6:6",
+            }.items()
         }
 
         # Tweak it to conform to the host architecture + update the version of a few dependencies
@@ -154,8 +164,8 @@ class ClingoBootstrapConcretizer:
                 node.versions = self.host_compiler.versions
 
             if node.name in to_be_updated:
-                pkg_cls = to_be_updated[node.name]
-                _select_best_version(pkg_cls=pkg_cls, node=node)
+                pkg_cls, valid_versions = to_be_updated[node.name]
+                _select_best_version(pkg_cls=pkg_cls, node=node, valid_versions=valid_versions)
 
         # Can't use re2c@3.1 with Python 3.6
         if self.host_python.satisfies("@3.6"):

--- a/lib/spack/spack/bootstrap/clingo.py
+++ b/lib/spack/spack/bootstrap/clingo.py
@@ -21,7 +21,6 @@ import spack.config
 import spack.package_base
 import spack.platforms
 import spack.repo
-import spack.solver.asp
 import spack.solver.versions
 import spack.spec
 import spack.traverse
@@ -194,7 +193,7 @@ class ClingoBootstrapConcretizer:
             if "libc" in edge.virtuals:
                 edge.spec = self.host_libc
 
-        spack.solver.asp._inject_patches_variant(s)
+        spack.spec._inject_patches_variant(s)
         s._finalize_concretization()
 
         # Work around the fact that the installer calls Spec.dependents() and

--- a/lib/spack/spack/bootstrap/clingo.py
+++ b/lib/spack/spack/bootstrap/clingo.py
@@ -21,6 +21,7 @@ import spack.config
 import spack.package_base
 import spack.platforms
 import spack.repo
+import spack.solver.asp
 import spack.solver.versions
 import spack.spec
 import spack.traverse
@@ -152,11 +153,17 @@ class ClingoBootstrapConcretizer:
                 "cmake": "3.16:3",
                 "libiconv": "1:1",
                 "ncurses": "6:6",
+                "m4": "1.4",
             }.items()
         }
 
         # Tweak it to conform to the host architecture + update the version of a few dependencies
         for node in s.traverse():
+            # Clear patches, we'll compute them correctly later
+            node.patches.clear()
+            if "patches" in node.variants:
+                del node.variants["patches"]
+
             node.architecture.os = str(self.host_os)
             node.architecture = self.host_architecture
 
@@ -187,6 +194,7 @@ class ClingoBootstrapConcretizer:
             if "libc" in edge.virtuals:
                 edge.spec = self.host_libc
 
+        spack.solver.asp._inject_patches_variant(s)
         s._finalize_concretization()
 
         # Work around the fact that the installer calls Spec.dependents() and

--- a/lib/spack/spack/package_base.py
+++ b/lib/spack/spack/package_base.py
@@ -2575,7 +2575,7 @@ def deprecated_version(pkg: PackageBase, version: Union[str, StandardVersion]) -
     return details is not None and details.get("deprecated", False)
 
 
-def preferred_version(pkg: PackageBase):
+def preferred_version(pkg: Union[PackageBase, Type[PackageBase]]):
     """
     Returns a sorted list of the preferred versions of the package.
 

--- a/lib/spack/spack/solver/asp.py
+++ b/lib/spack/spack/solver/asp.py
@@ -50,7 +50,6 @@ import spack.llnl.util.lang
 import spack.llnl.util.tty as tty
 import spack.package_base
 import spack.package_prefs
-import spack.patch
 import spack.platforms
 import spack.repo
 import spack.solver.splicing
@@ -3842,7 +3841,7 @@ class SpecBuilder:
         roots = [spec.root for spec in self._specs.values()]
         roots = dict((id(r), r) for r in roots)
         for root in roots.values():
-            _inject_patches_variant(root)
+            spack.spec._inject_patches_variant(root)
 
         # Add external paths to specs with just external modules
         for s in self._specs.values():
@@ -3954,79 +3953,6 @@ def _specs_with_commits(spec):
         " does not meet commit syntax requirements."
     )
     assert vn.is_git_commit_sha(spec.variants["commit"].value), invalid_commit_msg
-
-
-def _inject_patches_variant(root: spack.spec.Spec) -> None:
-    # This dictionary will store object IDs rather than Specs as keys
-    # since the Spec __hash__ will change as patches are added to them
-    spec_to_patches: Dict[int, Set[spack.patch.Patch]] = {}
-    for s in root.traverse():
-        # After concretizing, assign namespaces to anything left.
-        # Note that this doesn't count as a "change".  The repository
-        # configuration is constant throughout a spack run, and
-        # normalize and concretize evaluate Packages using Repo.get(),
-        # which respects precedence.  So, a namespace assignment isn't
-        # changing how a package name would have been interpreted and
-        # we can do it as late as possible to allow as much
-        # compatibility across repositories as possible.
-        if s.namespace is None:
-            s.namespace = spack.repo.PATH.repo_for_pkg(s.name).namespace
-
-        if s.concrete:
-            continue
-
-        # Add any patches from the package to the spec.
-        node_patches = {
-            patch
-            for cond, patch_list in spack.repo.PATH.get_pkg_class(s.fullname).patches.items()
-            if s.satisfies(cond)
-            for patch in patch_list
-        }
-        if node_patches:
-            spec_to_patches[id(s)] = node_patches
-
-    # Also record all patches required on dependencies by depends_on(..., patch=...)
-    for dspec in root.traverse_edges(deptype=dt.ALL, cover="edges", root=False):
-        if dspec.spec.concrete:
-            continue
-
-        pkg_deps = spack.repo.PATH.get_pkg_class(dspec.parent.fullname).dependencies
-
-        edge_patches: List[spack.patch.Patch] = []
-        for cond, deps_by_name in pkg_deps.items():
-            dependency = deps_by_name.get(dspec.spec.name)
-            if not dependency:
-                continue
-
-            if not dspec.parent.satisfies(cond):
-                continue
-
-            for pcond, patch_list in dependency.patches.items():
-                if dspec.spec.satisfies(pcond):
-                    edge_patches.extend(patch_list)
-
-        if edge_patches:
-            spec_to_patches.setdefault(id(dspec.spec), set()).update(edge_patches)
-
-    for spec in root.traverse():
-        if id(spec) not in spec_to_patches:
-            continue
-
-        patches = list(spec_to_patches[id(spec)])
-        variant: vt.VariantValue = spec.variants.setdefault(
-            "patches", vt.MultiValuedVariant("patches", ())
-        )
-        variant.set(*(p.sha256 for p in patches))
-        # FIXME: Monkey patches variant to store patches order
-        ordered_hashes = [(*p.ordering_key, p.sha256) for p in patches if p.ordering_key]
-        ordered_hashes.sort()
-        tty.debug(
-            f"Ordered hashes [{spec.name}]: "
-            + ", ".join("/".join(str(e) for e in t) for t in ordered_hashes)
-        )
-        setattr(
-            variant, "_patches_in_order_of_appearance", [sha256 for _, _, sha256 in ordered_hashes]
-        )
 
 
 def _ensure_external_path_if_external(spec: spack.spec.Spec) -> None:

--- a/lib/spack/spack/spec.py
+++ b/lib/spack/spack/spec.py
@@ -88,6 +88,7 @@ import spack.llnl.util.filesystem as fs
 import spack.llnl.util.lang as lang
 import spack.llnl.util.tty as tty
 import spack.llnl.util.tty.color as clr
+import spack.patch
 import spack.paths
 import spack.platforms
 import spack.provider_index
@@ -5572,6 +5573,79 @@ def eval_conditional(string):
     valid_variables = get_host_environment()
     valid_variables.update({"re": re, "env": os.environ})
     return eval(string, valid_variables)
+
+
+def _inject_patches_variant(root: Spec) -> None:
+    # This dictionary will store object IDs rather than Specs as keys
+    # since the Spec __hash__ will change as patches are added to them
+    spec_to_patches: Dict[int, Set[spack.patch.Patch]] = {}
+    for s in root.traverse():
+        # After concretizing, assign namespaces to anything left.
+        # Note that this doesn't count as a "change".  The repository
+        # configuration is constant throughout a spack run, and
+        # normalize and concretize evaluate Packages using Repo.get(),
+        # which respects precedence.  So, a namespace assignment isn't
+        # changing how a package name would have been interpreted and
+        # we can do it as late as possible to allow as much
+        # compatibility across repositories as possible.
+        if s.namespace is None:
+            s.namespace = spack.repo.PATH.repo_for_pkg(s.name).namespace
+
+        if s.concrete:
+            continue
+
+        # Add any patches from the package to the spec.
+        node_patches = {
+            patch
+            for cond, patch_list in spack.repo.PATH.get_pkg_class(s.fullname).patches.items()
+            if s.satisfies(cond)
+            for patch in patch_list
+        }
+        if node_patches:
+            spec_to_patches[id(s)] = node_patches
+
+    # Also record all patches required on dependencies by depends_on(..., patch=...)
+    for dspec in root.traverse_edges(deptype=dt.ALL, cover="edges", root=False):
+        if dspec.spec.concrete:
+            continue
+
+        pkg_deps = spack.repo.PATH.get_pkg_class(dspec.parent.fullname).dependencies
+
+        edge_patches: List[spack.patch.Patch] = []
+        for cond, deps_by_name in pkg_deps.items():
+            dependency = deps_by_name.get(dspec.spec.name)
+            if not dependency:
+                continue
+
+            if not dspec.parent.satisfies(cond):
+                continue
+
+            for pcond, patch_list in dependency.patches.items():
+                if dspec.spec.satisfies(pcond):
+                    edge_patches.extend(patch_list)
+
+        if edge_patches:
+            spec_to_patches.setdefault(id(dspec.spec), set()).update(edge_patches)
+
+    for spec in root.traverse():
+        if id(spec) not in spec_to_patches:
+            continue
+
+        patches = list(spec_to_patches[id(spec)])
+        variant: vt.VariantValue = spec.variants.setdefault(
+            "patches", vt.MultiValuedVariant("patches", ())
+        )
+        variant.set(*(p.sha256 for p in patches))
+        # FIXME: Monkey patches variant to store patches order
+        ordered_hashes = [(*p.ordering_key, p.sha256) for p in patches if p.ordering_key]
+        ordered_hashes.sort()
+        tty.debug(
+            f"Ordered hashes [{spec.name}]: "
+            + ", ".join("/".join(str(e) for e in t) for t in ordered_hashes)
+        )
+        setattr(
+            variant, "_patches_in_order_of_appearance", [sha256 for _, _, sha256 in ordered_hashes]
+        )
 
 
 class InvalidVariantForSpecError(spack.error.SpecError):

